### ISSUE TITLE
"ClientException: Not connected to voice" after attempting to play audio in a VoiceChannel

### ISSUE THREAD

**Issue by DavMrc** [Rapptz/discord.py#5995]
## Summary

My bot is always connected to a VoiceChannel and every time a user logs into such channel, I use the `on_voice_state_update` callback to play an audio in the channel.  
I noticed that sometimes (but not every time), the audio isn't played and I get this error: `ClientException: Not connected to voice`.  
I have tried to catch the exception, disconnect the bot for a second, then reconnect it and try to play the audio again, but doesn't solve the issue.

## Reproduction Steps

1. Create a bot
2. Register `on_voice_state_update`
3. Add some logic to filter only the case when a user is logging in/out
4. Play an audio
5. Host the bot on a 24/7 platform (don't think this matters much, but I was using Heroku. When I started to have this issue, I tried it also on my Raspberry Pi4)

## Expected Results

Bot always plays this audio.

## Actual Results

Sometimes (generally just an hour after the boot) I get this error and the bot doesn't play the audio. After that, generally everthing returns normal but it might happen again.

## Intents

My bot has all the Intents.

## Checklist

- [x] I have searched the open issues for duplicates.
- [x] I have provided the intents that my bot is using.
- [ ] I have shown the entire traceback, if possible.
- [ ] I have removed my token from display, if visible.

## System Information

When on Heroku, I know for sure that it utilizes Python 3.9.0 and the latest releases of `discord.py[voice]` and `aiohttp`. System is Ubuntu 18.04.

**Comment by Rapptz** [comment]
There is no code in this issue report making it inactionable. The reproduction steps provided are also too broad and vague to help with any form of debugging.

Please consider creating a minimal reproducible example that helps figure out the bug.

**Comment by DavMrc** [comment]
I have provided part of the `on_voice_state_update` callback, where the exception is raised.

**Comment by DavMrc** [comment]
> 
> 
> There is no code in this issue report making it inactionable. The reproduction steps provided are also too broad and vague to help with any form of debugging.
> 
> Please consider creating a minimal reproducible example that helps figure out the bug.

I have provided part of the `on_voice_state_update` callback, where the exception is raised.

**Comment by Gorialis** [comment]
Like most update events in the library, `on_voice_state_update` represents a larger scope than this.

As [documented](https://discordpy.readthedocs.io/en/latest/api.html#discord.on_voice_state_update), it occurs in these scenarios:
- **Any user** joining a voice channel
- **Any user** leaving a voice channel
- **Any user** muting/unmuting themselves
- **Any user** deafening/undeafening themselves
- **Any user** getting server muted or unmuted
- **Any user** getting server deafened or undeafened

As of newer updates, there are actually even more cases this occurs that aren't even listed here, including:
- **Any user** turning on or off their camera
- **Any user** starting or ending a Go-Live stream

Considering all of these factors, it makes sense why you would get exceptions about not being in a voice channel (in fact, the chance a voice state event will be specifically from your bot *and* pertain to joining a channel is quite slim in anywhere but a controlled environment).

You need to actually check the contents of the voice state updates you're receiving and compare the before and after states to figure out whether the event signals an appropriate time to start playing audio, if you wish to do it this way.

Honestly speaking though, since the only way that will occur on your end in a way that actually lets your code work as written, the join must be triggered through a [`VoiceChannel.connect`](https://discordpy.readthedocs.io/en/latest/api.html#discord.VoiceChannel.connect), which doesn't return until the connection is fully established anyway, so it would make more sense to just put your playing code alongside your connection code. This would also mean you wouldn't run into any issues later if you wished to write in other voice functionality features that required playing something else (or nothing at all).

**Comment by DavMrc** [comment]
> Considering all of these factors, it makes sense why you would get exceptions about not being in a voice channel (in fact, the chance a voice state event will be specifically from your bot and pertain to joining a channel is quite slim in anywhere but a controlled environment).
The bot doesn't trigger `on_voice_state_update`, **first** because I opt-out this case with an `if` case (I'll add it to my question since I figured I've mistakenly left that line out) and **second** because it always stays in the voice channel, never leaving.

> You need to actually check the contents of the voice state updates you're receiving and compare the before and after states to figure out whether the event signals an appropriate time to start playing audio, if you wish to do it this way.

This is something I already do, but since there are many cases that I'm filtering, I have considered not to include them in the code I provided. Just because the indentation of all the `if` and `elif` would make the code long and irrelevant for you to help me.

> Honestly speaking though, since the only way that will occur on your end in a way that actually lets your code work as written, the join must be triggered through a [`VoiceChannel.connect`](https://discordpy.readthedocs.io/en/latest/api.html#discord.VoiceChannel.connect), which doesn't return until the connection is fully established anyway, so it would make more sense to just put your playing code alongside your connection code. This would also mean you wouldn't run into any issues later if you wished to write in other voice functionality features that required playing something else (or nothing at all).

This part is not clear to me.  
I am guessing that maybe my code isn't clear enough, so I'll edit it.  
In short, I only want to play an audio when:
- an user connects to the channel
- an user disconnects from the channel
- an user disconnects from the server

**Comment by Gorialis** [comment]
> This is something I already do, but since there are many cases that I'm filtering, I have considered not to include them in the code I provided.

Please don't abridge code that you ask for help on. This has been the biggest bane of this issue - first you provide no code at all, and then you provide code that misleadingly leaves information out, obscuring your actual issue.

> ```python
>     if member.name != self.bot.user.name:  # opts out events triggered from the bot
> ```

If you need to compare users, just do so outright using `member != self.bot.user`. This compares ID, which is an absolute check, and will avoid events being blocked from other users that might have your bot's name.

> ```python
>         if str(voice_state_after.channel) == 'None':
> ```

Please use `voice_state_after.channel is None` for NoneType checks. Casting to string does nothing but create edge cases for you.

> ```python
>         elif (str(voice_state_before.channel) != str(voice_state_after.channel) and
>             str(voice_state_before.channel) == "General"  # name of the voice channel
>         ):
> ```

Like before, please don't just cast things to string for comparison. [`VoiceChannel`](https://discordpy.readthedocs.io/en/latest/api.html#discord.VoiceChannel) supports `__eq__`. The documentation even includes a table specifying implemented operators.
![image](https://user-images.githubusercontent.com/8049998/97810334-6e97cd80-1cb6-11eb-93c3-b7b09185fb3e.png)
Use this instead:
```python
elif voice_state_before.channel != voice_state_after.channel and voice_state_before.channel.name == 'General':
```


> This part is not clear to me.
> I am guessing that maybe my code isn't clear enough, so I'll edit it.
> In short, I only want to play an audio when:
> 
>     * an user connects to the channel
> 
>     * an user disconnects from the channel
> 
>     * an user disconnects from the server

This does help quite a lot.
Your code mentions that: ``# vc_connection was created when connecting via `await connect()` ``.
Since you're not getting AttributeError, it's most likely that this part is working fine, and that later on the `VoiceClient` becomes disconnected for whatever reason.

There's three main cases where this can occur:
1. The voice client was disconnected explicitly
2. The voice client is in the middle of an attempt to reconnect from a dropped socket
3. The voice client had the socket drop but had `reconnect` set to `False`

1 isn't very likely, and 3 isn't the default behavior, so the probable answer is 2 - the socket was dropped, either because the voice server crashed (they do this relatively often), or the internet connection between them was lost, and so you're "not connected to voice" because it's still busy trying to reconnect.

Since your events are temporal (that is - they're only relevant in the immediate moment following the event occurring), if it were me, I would just skip handling events during reconnects, since they'll probably constitute a relatively short amount of the overall connected time:
```python
if not self.vc_connection.is_connected():
    return
```

If that's not acceptable for whatever reason, you could use a Lock and force a manual reconnect within that Lock near the start of the function:

```python
# somewhere else, probably __init__
self.vc_lock = asyncio.Lock()



# within on_voice_state_update
async with self.vc_lock:  # this guarantees that if two voice state updates occur at the same time, the later-processed event won't disconnect the new manually connected client redundantly
    if not self.vc_connection.is_connected():
        await self.vc_connection.disconnect(force=True)
        self.vc_connection = await self.vc.connect()

...  # the rest
```
This approach has far more room for failure though, so I wouldn't really recommend it.

All-in-all, Discord's implementation for voice is kind of unreliable. If you want a 24/7 voice bot, you're going to have to do a lot of damage control to keep it working, since, well, Discord doesn't really expect users to remain connected for indefinite periods of time. Hopefully this might be of some use, though.

**Comment by DavMrc** [comment]
For my case, it would be best not to "miss" a single event, and therefore audio playback. Since the problem lies outside of the APIs (based on what you said) and since there is no way around it other than the harsh reconnect (which after testing, still doesn't play the audio for some reason) I'll consider dropping my attempt.  
Thanks anyway and pardon for the inconvenience.
